File: src/main/kotlin/supervinicius/util/Logger.java
package supervinicius.util;

import java.io.IOException;
import java.text.SimpleDateFormat;
import java.util.*;

public class Logger
{
    public enum LogTypes
    {
        COMMAND,
        DEBUG,
        WARNING,
        ERROR
    }

    private static final String RESET = "\u001B[0m";
    private static final String BLACK = "\u001B[30m";
    private static final String RED = "\u001B[31m";
    private static final String GREEN = "\u001B[32m";
    private static final String YELLOW = "\u001B[33m";
    private static final String BLUE = "\u001B[34m";
    private static final String PURPLE = "\u001B[35m";
    private static final String CYAN = "\u001B[36m";
    private static final String WHITE = "\u001B[37m";

    private static Logger instance;
    private String history;
    private final int MAX_SIZE = 10000;
    private Map<LogTypes, String> logMap = new HashMap<>();

    private Logger(){
        logMap.put(LogTypes.COMMAND, "Command");
        logMap.put(LogTypes.DEBUG, "Debug");
        logMap.put(LogTypes.WARNING, "Warn");
        logMap.put(LogTypes.ERROR, "Error");
    }

    public static Logger getInstance()
    {
        if(instance == null) instance = new Logger();
        return instance;
    }

    private void log(LogTypes logType, Object message, String color)
    {
        SimpleDateFormat formatter = new SimpleDateFormat("dd/MM/yyyy HH:mm:ss");
        Date date = new Date(System.currentTimeMillis());

        String logStr = "[" +
                formatter.format(date) +
                " | " +
                logMap.get(logType) +
                "] " +
                message.toString() +
                "\n";
        String print = color + logStr + RESET;

        System.out.print(print);

        history += logStr;
        if(history.length() == MAX_SIZE) saveLog();
    }

    public boolean saveLog(){
        SimpleDateFormat formatter = new SimpleDateFormat("dd-MM-yyyy HH-mm-ss");
        Date date = new Date(System.currentTimeMillis());
        boolean success = FileManager.getInstance()
                .writeToFile(history, "log"+ formatter.format(date) +".txt" , "logs");
        history = "";
        return success;
    }

    public void logCommand(Object message){ log(LogTypes.COMMAND, message, BLUE); }
    public void debug(Object message){ log(LogTypes.DEBUG, message, WHITE); }
    public void warn(Object message){ log(LogTypes.WARNING, message, YELLOW); }
    public void error(Object message){ log(LogTypes.ERROR, message, RED); }

}

File: src/main/kotlin/leaguebot/util/Logger.kt
package leaguebot.util

import java.text.SimpleDateFormat
import java.util.*

class Logger private constructor() {
    private var history: String = ""
    private val maxSize = 10000
    private val logMap = HashMap<LogTypes, String>()

    enum class LogTypes {
        COMMAND,
        DEBUG,
        WARNING,
        ERROR
    }

    init {
        logMap[LogTypes.COMMAND] = "Command"
        logMap[LogTypes.DEBUG] = "Debug"
        logMap[LogTypes.WARNING] = "Warn"
        logMap[LogTypes.ERROR] = "Error"
    }

    private fun log(logType: LogTypes, message: Any, color: String) {
        val formatter = SimpleDateFormat("dd/MM/yyyy HH:mm:ss")
        val date = Date(System.currentTimeMillis())

        val logStr = "[" +
                formatter.format(date) +
                " | " +
                logMap[logType] +
                "] " +
                message.toString() +
                "\n"
        val print = color + logStr + RESET

        print(print)

        history += logStr
        if (history.length == maxSize) saveLog()
    }

    fun saveLog(): Boolean {
        val formatter = SimpleDateFormat("dd-MM-yyyy HH-mm-ss")
        val date = Date(System.currentTimeMillis())
        val success = FileManager.getInstance()
            .writeToFile(history, "log" + formatter.format(date) + ".txt", "logs")
        history = ""
        return success
    }

    fun logCommand(message: Any) {
        log(LogTypes.COMMAND, message, BLUE)
    }

    fun debug(message: Any) {
        log(LogTypes.DEBUG, message, WHITE)
    }

    fun warn(message: Any) {
        log(LogTypes.WARNING, message, YELLOW)
    }

    fun error(message: Any) {
        log(LogTypes.ERROR, message, RED)
    }

    companion object {

        private val RESET = "\u001B[0m"
        private val BLACK = "\u001B[30m"
        private val RED = "\u001B[31m"
        private val GREEN = "\u001B[32m"
        private val YELLOW = "\u001B[33m"
        private val BLUE = "\u001B[34m"
        private val PURPLE = "\u001B[35m"
        private val CYAN = "\u001B[36m"
        private val WHITE = "\u001B[37m"

        private var instance: Logger? = null

        fun getInstance(): Logger {
            if (instance == null) instance = Logger()
            return instance as Logger
        }
    }

}
